### PR TITLE
Implemented Cascade DDS push API - Derecho Notification Mechanism

### PR DESCRIPTION
**Pull request by ZhengPeterWang**
Implemented the notification mechanism of derecho. Servers are now able to use this API to send external clients messages. Please try src/applications/archive/notification_test.cpp for more details.

### CODE REVIEW

**Comment by ZhengPeterWang** [comment]
Hi Edward,
I noticed that there is another merge problem introduced by #221. In this merge, src/core/view_manager.cpp was also obsolete. The 16 lines of put_with_completion() should all be put(). These are not my intended changes and should be removed from derecho's client_callback branch.
Thank you!
Best,
Zheng Wang

**Comment by KenBirman** [comment]
Peter, the choice between put with completion and put is a tricky one.  Sagar should review this merge if there is any possibility that you are looking at edits he made.  This is actually tied to Weijia’s worry about people breaking the main branch.  Put with completion is quite different from put, and Sagar’s code has very delicate dependencies on whether or not a completion is requested.  Edward’s code similarly has a lot of subtle interactions with RDMA completions.

Basically, a put with completion is needed in situations where the sender of some data has to be certain the data reached some destination or set of destinations.  It is like waiting for an ack.  And of course that is VERY different than not waiting for the ack, which is what put (no completion) gives us.

So this dialog between you and Edward suddenly has me very anxious!

**Comment by etremel** [comment]
It took me a while to find this discussion because it's on a finished (closed) pull request, but I think I can answer your questions about the view_manager.cpp changes. In the current master branch of Derecho, there are several lines in view_manager.cpp that are currently `put()`, and were changed from `put_with_completion()` to `put()` by this commit: 9a57a9bd4fb4ab95fc9af07d6c892c153761823e. That commit is also present in your client_callbacks branch, so it hasn't been "lost." The reason why the comparison between client_callbacks and master shows these lines changing from `put()` back to `put_with_completion()` is because you changed them yourself, in the client_callbacks branch, in this commit: 158dba18f237ff81794716f7907bcc6452cff52e.

Did you intend to change the `put()` calls to `put_with_completion()` when you were working on client_callbacks? If so, did you consult with Sagar about it?

**Comment by ZhengPeterWang** [comment]
I did not intend to change the` put()` calls to `put_with_completion()`. That change was, once again, introduced because I was working on an old version of derecho from years ago for the client_callbacks branch. I haven't consulted Sagar on this issue yet, but for safety, I think we should roll back to `put()` as indicated in `9a57a9b`.
